packages/react-dom/src/events/replayedEvent.js

```diff
@@ -0,0 +1,23 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+import type {AnyNativeEvent} from '../events/PluginModuleType';
+
+const eventsReplaying = new Set();
+
+// This exists to avoid circular dependency between ReactDOMEventReplaying
+// and DOMPluginEventSystem
+export function replayEventWrapper(event: AnyNativeEvent, replay: () => void) {
+  eventsReplaying.add(event);
+  replay();
```

Comment:
Should we be a bit more protective here and use try/finally? Though we don't expect the things we pass today to `replay` could throw, might be safer in longer term.

Also is it worth protecting against reentrancy?

```js
  const wasReplaying = eventsReplaying.has(event);
  eventsReplaying.add(event);
  try {
    replay();
  } finally {
    if (wasReplaying) {
      eventsReplaying.delete(event);
    }
  }
```

so only the outermost counts

Reply:
I don't think its possible for this to be reentrant and if it were I'm not sure that blocking the replaying would be the right behavior either. Maybe I can add an invariant here?

Reply:
For it to be re-entrant, the event would need to be manually re-dispatched in userland, while the event is already being dispatched. However, when the user re-dispatches it we'll see that it's not blocked on anything so it wouldn't go through this codepath. It would just be directly dispatched.

Reply:
If we're super confident in this we can add an invariant.